src/RateLimiter/RateLimiterMiddleware.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * RateLimiter helps to prevent abuse by limiting the number of requests that could be me made consequentially.
+ *
+ * For example, you may want to limit the API usage of each user to be at most 100 API calls within a period of 10 minutes.
+ * If too many requests are received from a user within the stated period of the time, a response with status code 429
+ * (meaning "Too Many Requests") should be returned.
+ */
+final class RateLimiterMiddleware implements MiddlewareInterface
+{
+    private CounterInterface $counter;
+
+    private ResponseFactoryInterface $responseFactory;
+
+    private string $counterId;
+
+    /**
+     * @var callable
+     */
+    private $counterIdCallback;
+
+    public function __construct(CounterInterface $counter, ResponseFactoryInterface $responseFactory)
+    {
+        $this->counter = $counter;
+        $this->responseFactory = $responseFactory;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $this->counter->setId($this->generateId($request));
+        $result = $this->counter->incrementAndGetResult();
+
+        if ($result->isLimitReached()) {
+            $response = $this->createErrorResponse();
+        } else {
+            $response = $handler->handle($request);
```

Comment:
If this section runs for a long time, the value of `$ result` will be outdated.

Reply:
Method getReset now returns the time.